File: algo_problems/q700+/q763.py
class Solution:
    def partitionLabels(self, S):
        """
        :type S: str
        :rtype: List[int]
        """
        st = set()
        startList = [' ' for i in range(len(S))]
        for i, c in enumerate(S):
            if c not in st:
                st.add(c)
                startList[i] = c

        st = set()
        endList = [' ' for i in range(len(S))]
        for i, c in enumerate(S[::-1]):
            if c not in st:
                st.add(c)
                endList[i] = c
        endList = endList[::-1]

        cnt = 0
        preIndex = 0
        result = []
        for i in range(len(S)):
            if startList[i] != ' ':
                cnt += 1
            if endList[i] != ' ':
                cnt -= 1
            if cnt == 0:
                result.append(i + 1 - preIndex)
                preIndex = i + 1
        if preIndex < len(S):
            result.append(len(S) - preIndex)
        return result


if __name__ == '__main__':
    print(Solution().partitionLabels('ababcbacadefegdehijhklij'))

File: algo_problems/q700+/q763.java
class Solution {
    public List<Integer> partitionLabels(String S) {
        Set<Character> hs = new HashSet();
        ArrayList<Character> startIndex = new ArrayList();
        for (int i = 0; i < S.length(); i++) {
        	char c = S.charAt(i);
        	if (!hs.contains(c)) {
        		hs.add(c);
        		startIndex.add(c);
        	}else{
        		startIndex.add(' ');
        	}
        }
        ArrayList<Character> endIndex = new ArrayList();
        hs = new HashSet();
        for (int i = S.length() - 1; i > -1; i--) {
        	char c = S.charAt(i);
        	if (!hs.contains(c)) {
        		hs.add(c);
        		endIndex.add(0, c);
        	}else{
        		endIndex.add(0, ' ');
        	}
        }

        int cnt = 0;
        int preIndex = 0;
        List<Integer> result = new ArrayList();
        for (int i = 0; i < S.length(); i++){
        	if (startIndex.get(i) != ' ') {
        		cnt += 1;
        	}
        	if (endIndex.get(i) != ' ') {
        		cnt -= 1;
        	}
        	if (cnt == 0) {
        		result.add(i + 1 - preIndex);
                preIndex = i + 1;
        	}
        }
        return result;
    }
}